Posting comment on issue (Feature/28)

Declow0 opened the pull request:
Данная ветка добавляет пустой конструктор к Comment.java и вносит некоторые изменения в подрядок функций для корректной работы с JSON. Создан класс CommentController. Его end-point "/postcomment". Изменены в соответствии html и js файлы для постинга комментариев. Добавлена зависимость для работы с LocalDateTime как JSON.

Card: #28

Commit:
Posting comment on issue

Данная ветка добавляет пустой конструктор к Comment.java и вносит некоторые изменения в подрядок функций для корректной работы с JSON. Создан класс CommentController. Его end-point "/postcomment". Изменены в соответствии html и js файлы для постинга комментариев. Добавлена зависимость для работы с LocalDateTime как JSON.

Card: #28

Declow0 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Как разраб фронтенда, хочу эндпойнт для сущности "Комментарий", чтобы мой фронтенд мог POSTить комменты](https://trello.com/c/TWOA9zxz/28-post)

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java:
```diff
@@ -5,33 +5,42 @@
+    public Comment() {
+        this.author = "";
+        this.content = "";
+        createdDateTime = null;
```
А зачем здесь без `this`? Выглядит так, что можно и остальное без `this`. Или вообще унести в инициализацию.

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java:
```diff
@@ -5,33 +5,42 @@
-    public Comment(String author, String content) {
-        createdDateTime = LocalDateTime.now();
+    public Comment() {
+        this.author = "";
```
А для чего `author` пустая строка? Почему бы не держать по умолчанию `null`? (это не призыв заменить, это призыв разъяснить)

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/CommentController.java:
```diff
@@ -0,0 +1,14 @@
+
+import org.springframework.web.bind.annotation.*;
+
+@RestController
```
Как обсудили в чатике, будет лучше путь `/issues/{issueHash}/comments`

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/CommentController.java:
```diff
@@ -0,0 +1,14 @@
+
+@RestController
+public class CommentController {
+    @RequestMapping(value = "/postcomment", method = RequestMethod.POST)
```
Есть алиас `@PostMapping`, который являет собой `@RequestMapping` с методом POST. В других местах используем эти алиасы, давайте придерживаться этого по всему проекту.

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/CommentController.java:
```diff
@@ -0,0 +1,14 @@
+@RestController
+public class CommentController {
+    @RequestMapping(value = "/postcomment", method = RequestMethod.POST)
+    public void post(@RequestBody Comment comment, @RequestParam("issueHash") long issueHash) {
```
`@RequestParam` это же который после знака вопроса в URL? Давай лучше `@PathParam`, или как там в спринге.

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/resources/static/js/postcomment.js:
```diff
@@ -1,12 +1,17 @@
+			url	:	'/postcomment',
+			data	:	$scope.comment,
+			params  :   {issueHash: $scope.issueHash},
+			headers :   {'Content-Type': 'application/json'}
```
А по умолчанию не ставит `Content-Type`? Вот зараза...

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/resources/static/js/postcomment.js:
```diff
@@ -1,12 +1,17 @@
 	$scope.submitForm = function() {
+	    if ($scope.comment.author == null) {
+                $scope.comment.author = "Anonymous";
+        }
```
Тут какая-то печаль с выравниванием кода, где-то табы, где-то пробелы. Давай везде использовать что-то одно.

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/resources/static/js/postcomment.js:
```diff
@@ -1,12 +1,17 @@
-	$scope.comment = {};
+    $scope.comment = {};
 	$scope.submitForm = function() {
+	    if ($scope.comment.author == null) {
```
Может быть ещё `undefined` и пустая строка. Лучше `if (!$scope.comment.author)`.

Declow0 replied:
Спасибо, не знал что тут писать, ибо ничего не работало.

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java:
```diff
@@ -5,33 +5,42 @@
+    }
+
+    public void setCreatedDateTime() {
+        this.createdDateTime = LocalDateTime.now();
```
Вот это очень странно, что setter без параметра. Как тут поступить правильно - не знаю, но явно не так :) Сеттер это требование десериализации?

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/pom.xml:
```diff
@@ -25,6 +25,12 @@
       <optional>true</optional>
     </dependency>
+	
+	<dependency>
```
Здесь тоже какая-то печаль с выравниванием. Отровняй плз всю pom'ку в едином стиле.

Declow0 replied:
Done.

Commit:
Исправлены ошибки форматирования и исправлены недочеты

Пустой конструктор Comment заполнен null, избавлен от this. Метод setCreatedDateTime() переименован в setCurrentTime(). Изменен путь к end-point, изменена аннотация на @PostMapping. @RequestParam изменена на @PathParam. Изменено условие проверки пустоты автора if (!$scope.comment.author). Content-Type убран, так как является значением по умолчанию. Исправлены выравнивания.

Card: #28

Commit:
Change name of variable hash to issueHash

Card: #28

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java:
```diff
@@ -5,33 +5,42 @@
 
-    public Comment(String author, String content) {
-        createdDateTime = LocalDateTime.now();
+    public Comment() {
```
Нам точно нужен пустой конструктор?
Если даже да - приравнивать к null каждое поле класса в нём не нужно, это автоматически делается.

Declow0 replied:
Да, для @RequestBody необходим. Ну вообще, да по спецификации Java все неинициализированные поля равны null.

Declow0 replied:
Done.

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java:
```diff
@@ -5,33 +5,42 @@
+    public Comment() {
+        author = null;
+        content = null;
+        createdDateTime = null;
```
предлагаю now() - что такое нулевая дата?

Declow0 replied:
Не инициализированная, но now() установлю.

Declow0 replied:
Done. Так же это помогает полям не инициализированным в JSON, инициализироваться по данному конструктору.

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java:
```diff
@@ -5,33 +5,42 @@
+        this.createdDateTime = LocalDateTime.now();
     }
 
-    public String getAuthor() {
```
Переколбашены методы - ок, пусть будет, но тут стоит учитывать возможные проблемы мёрджа этой ветки с другими изменениями. Такие переколбасы стоит делать только когда они действительно нужны и полезны.

Declow0 replied:
Я их делал, потому что с JSON проблемы были, но видимо зря. Done.

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java:
```diff
@@ -5,33 +5,42 @@
+        this.createdDateTime = LocalDateTime.now();
     }
-}
+}
```
пустая строка в конце файла

Declow0 replied:
Done.

kevromster commented on backend/tracker-web-app.backend/src/main/resources/static/js/postcomment.js:
```diff
@@ -3,10 +3,13 @@ postApp.controller('postController', function($scope, $http)
 		$http({
 			method	:	'POST',
-			url	:	'postcomment',
+			url	:	'/issues/'+$scope.issueHash+'/comments',
```
операторы лучше пробелами отделять: x + y вместо x+y

Declow0 replied:
Done.

kevromster commented on backend/tracker-web-app.backend/src/main/resources/static/js/postcomment.js:
```diff
@@ -3,10 +3,13 @@ postApp.controller('postController', function($scope, $http)
-		}).success(function() {});
+		});
 	};
 });
```
пустая строка в конце файла

Declow0 replied:
Done.

mgtriffid commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/CommentController.java:
```diff
@@ -0,0 +1,14 @@
+
+@RestController
+public class CommentController {
+    @PostMapping(value = "/issues/{issueHash}/comments")
```
Этот путь будет общим для комментов, так что его можно повешать на класс, а на методы потом будем вешать просто хвостики пути, типа `/{hash}`.

Declow0 replied:
Done. Путь указывается в @RestMapping.

Commit:
Change contrustor Comment(). Add '\n' at end of some files.

Card: #28

Commit:
Delete setCurrentTime(), because it sets in Comment() constructor. Path in @RequestMapping

Card: #28

Commit:
Alerting then fields Number of issue and comment are empty

Card: #28

Commit:
Add status of comment and you can see some errors

Card: #28

Commit:
Add '\n' at the end to postcomment.html

Card: #28

Declow0 commented:
А можно пока не принимать pull request, поскольку хочется проверить возможность заполнять анонимного автора не в JS, а через конструктор без параметров.

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/CommentController.java:
```diff
@@ -0,0 +1,14 @@
+public class CommentController {
+    @PostMapping
+    public void post(@RequestBody Comment comment, @PathVariable("issueHash") long hash) {
+        IssueController issueController = new IssueController();
```
стоп, мы тут зачем создаём объект _IssueController_? Чтобы получить доступ до статической мапы? А откуда мы тут знаем, что там статическая мапа? Не, не пойдёт

Declow0 replied:
Сделал getIssue статик методом, больше ничего не придумал пока. Done.

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/CommentController.java:
```diff
@@ -0,0 +1,14 @@
+    @PostMapping
+    public void post(@RequestBody Comment comment, @PathVariable("issueHash") long hash) {
+        IssueController issueController = new IssueController();
+        Issue issue = issueController.getIssue(hash);
```
а если нет такой ишуи, что произойдёт?

Declow0 replied:
Ошибка 400/404 не помню точно, что его нет.

kevromster commented on backend/tracker-web-app.backend/src/main/resources/static/js/postcomment.js:
```diff
@@ -3,10 +3,27 @@ postApp.controller('postController', function($scope, $http)
 			data	:	$scope.comment
-		}).success(function() {});
+		})
+			.success(function() {
```
тут зачем-то отступ увеличил, имхо некрасиво

Declow0 replied:
http://contribute.jquery.org/style-guide/js/
Этого требует официальный стиль Chained Method Calls.

Также можно найти именно для Angular: Defer Controller Logic to Services.
https://github.com/johnpapa/angular-styleguide/blob/master/a1/README.md

Просят избегать случаев, когда нет отступа.

mgtriffid replied:
Имхо норм с отступами

kevromster replied:
ok

kevromster commented on backend/tracker-web-app.backend/src/main/resources/static/js/postcomment.js:
```diff
@@ -3,10 +3,27 @@ postApp.controller('postController', function($scope, $http)
+				$scope.status = 'Comment posted.';
+			})
+			.error(function() {
+				$scope.status = 'Comment didn\'t poste.';
```
post вместо poste

Declow0 replied:
Done.

kevromster commented on backend/tracker-web-app.backend/src/main/resources/static/js/postcomment.js:
```diff
@@ -3,10 +3,27 @@ postApp.controller('postController', function($scope, $http)
 {
 	$scope.comment = {};
 	$scope.submitForm = function() {
+		if ($scope.issueHash == undefined) {
```
а какого типа тут будет реально issueHash? число 3 или строка '3'?

Declow0 replied:
В html файле явно прописано type="number". Строки быть не может, поскольку в итоге должно быть поле формата long. А поскольку hash может быть равен 0, так и написал. так как !0=1 и код считает, что есть ошибка.

kevromster replied:
Вообще, в javascript всё очень хитро с проверками на undefined. Может, здесь и ок, не знаю, но погугли на тему javascript check for undefined. Возможно, @mgtriffid тут больше скажет.

Declow0 replied:
http://www.codereadability.com/how-to-check-for-undefined-in-javascript/
Вот, что я нашел. Короче написано правильно, но если будет браузер IE 8 или старше, то вроде не сработает. Но данное написание удовлетворяет стандарту ECMAScript 5 2009 года.

kevromster replied:
ok

Commit:
Put author in Comment() constructor

Card: #28

Commit:
Fixed error in word "post"

Card: #28

Commit:
getIssue() now static method

Card: #28

Commit:
Change import from annotation.* to detailed and change name of method post

Card: #28

Commit:
Delete all set methods from Comment

Card: #28

Commit:
Add setContent for changing comment

Card: #28

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/IssueController.java:
```diff
@@ -52,7 +52,7 @@
 
     @GetMapping(value = "/{hash}")
-    public Issue getIssue(@PathVariable("hash") long hash) {
+    public static Issue getIssue(@PathVariable("hash") long hash) {
```
_static_ не пойдёт - тут либо выносить хранилище ишуёв в отдельный сервис, либо то, что я предложил в _CommentController_

Declow0 replied:
Больше не static, done.

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/CommentController.java:
```diff
@@ -0,0 +1,18 @@
+@RequestMapping(value = "/issues/{issueHash}/comments")
+public class CommentController {
+    @PostMapping
+    public void postComment(@RequestBody Comment comment, @PathVariable("issueHash") long hash) {
```
Что-то не соображу, а зачем нам сейчас вообще отдельный контроллер для комментов? В таком вот виде - почему этот _post_ не сделать частью _IssueController_? Чем добавление комментария тут отличается от обновления дескрипшна ишуя? Тогда и со статическим доступом к хранилищу проблем не будет.

Declow0 replied:
PostComment теперь часть IsssueController. Done

Commit:
Now getIssue is not-static. Delete class Comment controller

Now postComment is a part of IssueController.

Card: #28

Commit:
Change sequence of methods in Comment

Card: #28

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java:
```diff
@@ -8,10 +8,14 @@
 
-    public Comment(String author, String content) {
+    public Comment() {
+        author = "Anonymous";
```
можно сэкономить код, позвав тут this("Anonymous", null);

Declow0 replied:
Done.

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/IssueController.java:
```diff
@@ -60,6 +60,12 @@ public Issue getIssue(@PathVariable("hash") long hash) {
 
+    @PostMapping(value = "{issueHash}/comments")
+    public void postComment(@RequestBody Comment comment, @PathVariable("issueHash") long hash) {
+        Issue issue = issueMap.get(hash);
```
а если такого ишуя нет? вернётся null, и в строке ниже будет NPE

Declow0 replied:
Done.

Declow0 committed:
This branch has conflicts

Card: #28

kevromster commented on backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/IssueController.java:
```diff
@@ -74,4 +74,10 @@ public Issue getIssue(@PathVariable("hash") long hash) {
+    @PostMapping(value = "{issueHash}/comments")
+    public void postComment(@RequestBody Comment comment, @PathVariable("issueHash") long hash) {
+        Issue issue = issueMap.get(hash);
+        issue.addComment(comment);
```
нужна проверка на null

Declow0 committed:
Merge branch 'master' into feature/28

Conflicts:
	backend/tracker-web-app.backend/pom.xml
	backend/tracker-web-app.backend/src/main/java/edu/xored/tracker/Comment.java

Card: #28

Declow0 committed:
Change constructor of Comment(), check for null hash map

Card: #28

kevromster reviewed (approved):
LGTM

Declow0 committed:
Change name of PathVariable from "issueHash" to "hash"

Card: #28

mgtriffid reviewed (approved):
Мержабельно.
